kaolin/datasets/base.py

```diff
@@ -0,0 +1,154 @@
+        name = get_attributes(idx)['name']
+        if name not in cache_transform.cached_ids:
+            data = get_data(idx)
+            cache_transform(name, *data)
+
+class KaolinDatasetMeta(type):
+    def __new__(metacls, cls_name, base_cls, class_dict):
+        if cls_name != "KaolinDataset":
+            class_dict['__doc__'] += \
+                """Additional args:
+        preprocessing_params (dict): parameters for the preprocessing:
+            - 'cache_dir': path to the cached preprocessed data.
+            - 'num_workers': number of process used in parallel for preprocessing (default: number of cores)
+        preprocessing_transform (Callable): Called on the outputs of _get_data over the indices
+                                            from 0 to len(self) during the construction of the dataset,
+                                            the preprocessed outputs are then cached to 'cache_dir'.
+        transform (Callable): Called on the preprocessed data at __getitem__.
+        no_progress (bool): disable tqdm progress bar for preprocessing."""
+        return type.__new__(metacls, cls_name, base_cls, class_dict)
+
+class KaolinDataset(Dataset, metaclass=KaolinDatasetMeta):
+    """
+    Abstract class for dataset with handling of multiprocess or cuda preprocessing.
+
+    A KaolinDataset children class will need the above implementation:
+       1) _initialize:
+           Initialization function called at the beginning of the constructor.
+       2) _get_data:
+           Data getter that will be preprocessed => cached => transformed, take an index as input.
+       3) _get_attributes:
+           Attributes getter that will be preprocess / transform independent.
+       4) __len__:
+           Return the size of the dataset
+    """
+    def __init__(self, *args, preprocessing_transform=None, preprocessing_params: dict = None,
+                 transform=None, no_progress: bool = False, **kwargs):
+        """
+        Args:
+            positional and keyword arguments for initialize(*args, **kwargs) (see class and initialize documentation)
+            preprocessing_params (dict): parameters for the preprocessing:
+                - 'cache_dir': path to the cached preprocessed data.
+                - 'num_workers': number of process used in parallel for preprocessing (default: number of cores)
+            preprocessing_transform (Callable): Called on the outputs of _get_data over the indices
+                                                from 0 to len(self) during the construction of the dataset,
+                                                the preprocessed outputs are then cached to 'cache_dir'.
+            transform (Callable): Called on the preprocessed data at __getitem__.
+            no_progress (bool): disable tqdm progress bar for preprocessing.
+        """
+        self.initialize(*args, **kwargs)
+        if preprocessing_transform is not None:
+            desc = 'applying preprocessing'
```

Comment:
'applying prepreocessing' -> 'preprocessing ...'?

Reply:
? It's already 'applying preprocessing'

Reply:
Yeah, I was suggesting changing to just `preprocessing ...`.

Reply:
I think it's more descriptive to have:
```applying preprocessing: 100%|##################################| 75/75 [00:00<00:00, 1893.97it/s]```

Reply:
`Preprocessing` is a verb in itself, so I'm leaning towards it. But this is fine too, if you prefer.